akd_local_auditor/src/storage/s3/mod.rs

```diff
@@ -69,22 +105,47 @@ impl From<&S3ClapSettings> for S3AuditStorage {
     fn from(clap: &S3ClapSettings) -> Self {
         Self {
             bucket: clap.bucket.clone(),
+            region: clap.region.clone(),
+            endpoint: clap.endpoint.as_ref().cloned(),
+            access_key: clap.access_key.as_ref().cloned(),
+            secret_key: clap.secret_key.as_ref().cloned(),
             config: Arc::new(RwLock::new(None)),
             cache: Arc::new(RwLock::new(None)),
         }
     }
 }
 
 impl S3AuditStorage {
-    async fn get_config(&self) -> s3::Config {
+    async fn get_shared_config(&self) -> aws_types::SdkConfig {
         let mut lock = self.config.write().await;
-        let shared_config = if let Some(config) = &*lock {
+        if let Some(config) = &*lock {
             config.clone()
         } else {
-            let sc = aws_config::load_from_env().await;
-            *lock = Some(sc.clone());
-            sc
-        };
+            // Get the shared AWS config
+            let region_provider = RegionProviderChain::first_try(Region::new(self.region.clone()));
+
+            let mut shared_config_loader = aws_config::from_env().region(region_provider);
+
+            if let (Some(access_key), Some(secret_key)) = (&self.access_key, &self.secret_key) {
+                let credentials = aws_types::Credentials::from_keys(access_key, secret_key, None);
+                shared_config_loader = shared_config_loader.credentials_provider(credentials);
+            }
+
+            if let Some(endpoint) = &self.endpoint {
+                // THE URI IS ALREADY VALIDATED BY CLAP ARGS, hence the expect here
+                let endpoint_resolver = Endpoint::immutable(endpoint.parse().expect("valid URI"));
```

Comment:
Likely a good expect. It looks like URI might allow relative URIs, which I don't _think_ you want here based on usage. See: https://docs.rs/http/latest/http/uri/struct.Uri.html#examples-5.

Reply:
Good point, I'm updating the parser to reflect relative uris not being supported